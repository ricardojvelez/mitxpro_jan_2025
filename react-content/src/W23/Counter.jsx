import React, { useState } from 'react';

const Counter = () => {
  const [count, setCount] = useState(0); // count is the state variable, setCount is the function to update the state

  const increment = () => {
    setCount(count + 1);
  };

  const decrement = () => {
    setCount(count - 1);
  };

  return (
    <div>
      <h2>App Counter</h2>
      <p>Count: {count}</p>
      <button style={{marginRight: "8px"}} onClick={increment}>Increment</button>
      <button onClick={decrement}>Decrement</button>
    </div>
  );
}

export default Counter;