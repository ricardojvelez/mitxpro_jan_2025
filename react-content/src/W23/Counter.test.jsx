import React from "react";
import userEvent from "@testing-library/user-event";
import { render, screen } from "@testing-library/react";
import Counter from "./Counter";

describe("Counter", () => {
  it("renders the component", () => {
    render(<Counter />);

    const title = screen.getByText("App Counter");
    const countParagraph = screen.getByText("Count: 0");
    const incrementButton = screen.getByText("Increment");
    const decrementButton = screen.getByText("Decrement");

    expect(title).toBeInTheDocument();
    expect(countParagraph).toBeInTheDocument();
    expect(incrementButton).toBeInTheDocument();
    expect(decrementButton).toBeInTheDocument();

    //Another way to test
    //expect(screen.getByText("App Counterrr")).toBeInTheDocument();
  });

  it("renders the component with snapshot", () => {
    const { container } = render(<Counter />);

    expect(container).toMatchSnapshot();
  });

  describe("when user clicks Increment", () => {
    it("should render the count with an incremented value", () => {
      render(<Counter />);

      const incrementButton = screen.getByRole("button", { name: "Increment" });
      userEvent.click(incrementButton);

      expect(screen.getByText("Count: 1")).toBeInTheDocument();
    });
  });

  describe("when user clicks Decrement", () => {
    it("should render the count with a decremented value", () => {
      render(<Counter />);

      //Not so optimal way to find because it's not verifying the role!!! Should use getByRole ;)
      const decrementButton = screen.getByText("Decrement"); 
      userEvent.click(decrementButton);

      expect(screen.getByText("Count: -1")).toBeInTheDocument();
    });
  });
});

