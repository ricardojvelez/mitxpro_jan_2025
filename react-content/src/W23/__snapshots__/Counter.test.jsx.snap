// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Counter renders the component with snapshot 1`] = `
<div>
  <div>
    <h2>
      App Counter
    </h2>
    <p>
      Count: 
      0
    </p>
    <button
      style="margin-right: 8px;"
    >
      Increment
    </button>
    <button>
      Decrement
    </button>
  </div>
</div>
`;
